diff --git a/flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/StreamingFileWriter.java b/flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/StreamingFileWriter.java
index c2186bff9bc..924ef982103 100644
--- a/flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/StreamingFileWriter.java
+++ b/flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/StreamingFileWriter.java
@@ -79,17 +79,10 @@ public class StreamingFileWriter extends AbstractStreamOperator<CommitMessage>
 	public void initializeState(StateInitializationContext context) throws Exception {
 		super.initializeState(context);
 		buckets = bucketsBuilder.createBuckets(getRuntimeContext().getIndexOfThisSubtask());
-		helper = new StreamingFileSinkHelper<>(
-				buckets,
-				context.isRestored(),
-				context.getOperatorStateStore(),
-				getRuntimeContext().getProcessingTimeService(),
-				bucketCheckInterval);
 
+		// Set listener before the initialization of Buckets.
 		inactivePartitions = new HashSet<>();
-		currentWatermark = Long.MIN_VALUE;
 		buckets.setBucketLifeCycleListener(new BucketLifeCycleListener<RowData, String>() {
-
 			@Override
 			public void bucketCreated(Bucket<RowData, String> bucket) {
 			}
@@ -99,6 +92,14 @@ public class StreamingFileWriter extends AbstractStreamOperator<CommitMessage>
 				inactivePartitions.add(bucket.getBucketId());
 			}
 		});
+
+		helper = new StreamingFileSinkHelper<>(
+				buckets,
+				context.isRestored(),
+				context.getOperatorStateStore(),
+				getRuntimeContext().getProcessingTimeService(),
+				bucketCheckInterval);
+		currentWatermark = Long.MIN_VALUE;
 	}
 
 	@Override
