diff --git a/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java b/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java
index 581f24c8ec7..df289b5b98a 100644
--- a/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java
+++ b/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java
@@ -89,7 +89,7 @@ public class AppendOnlyFirstNFunction extends AbstractTopNFunction {
         currentRank += 1;
         state.update(currentRank);
 
-        if (outputRankNumber) {
+        if (outputRankNumber || hasOffset()) {
             collectInsert(out, input, currentRank);
         } else {
             collectInsert(out, input);
diff --git a/flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunctionTest.java b/flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunctionTest.java
index 85f80dd5a73..914cbc68da8 100644
--- a/flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunctionTest.java
+++ b/flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunctionTest.java
@@ -135,9 +135,7 @@ public class AppendOnlyFirstNFunctionTest extends TopNFunctionTestBase {
         testHarness.close();
 
         List<Object> expectedOutput = new ArrayList<>();
-        expectedOutput.add(insertRecord("book", 2L, 12));
         expectedOutput.add(insertRecord("book", 2L, 19));
-        expectedOutput.add(insertRecord("fruit", 1L, 33));
         expectedOutput.add(insertRecord("fruit", 1L, 44));
         assertorWithRowNumber.assertOutputEquals(
                 "output wrong.", expectedOutput, testHarness.getOutput());
